index.js

```diff
@@ -22,28 +22,42 @@ var express = require("express");
 var http = require('http');
 var stoppable = require('stoppable');
 
+var RED;
+var redNodes;
+var flows;
+var credentials;
+var comms;
+var log;
+var context;
+var events;
+
+function initRuntime(requirePath) {
+
+    try {
+        RED = require(requirePath);
+        var prefix = requirePath.substring(0, requirePath.indexOf('/red.js'));
+        redNodes = require(prefix+"/runtime/nodes");
+        flows = require(prefix+"/runtime/nodes/flows");
+        credentials = require(prefix+"/runtime/nodes/credentials");
+        comms = require(prefix+"/api/editor/comms");
+        log = require(prefix+"/runtime/log");
+        context = require(prefix+"/runtime/nodes/context");
+        events = require(prefix+"/runtime/events");
+    } catch (err) {
+        // ignore, assume init will be called again by a test script supplying the runtime path
+    }
+}
+
+// assume we have node red as a dependency
 try {
-    var RED = require('node-red');
-    var redNodes = require("node-red/red/runtime/nodes");
-    var flows = require("node-red/red/runtime/nodes/flows");
-    var credentials = require("node-red/red/runtime/nodes/credentials");
-    var comms = require("node-red/red/api/editor/comms.js");
-    var log = require("node-red/red/runtime/log.js");
-    var context = require("node-red/red/runtime/nodes/context.js");
-    var events = require('node-red/red/runtime/events');
+    var runtimePath = require.resolve('node-red');
 } catch (err) {
-    // no node-red in helper-test dependencies so assume we're testing node-red
-    var nrPath = process.cwd();
-    var RED = require(nrPath+"/red/red.js");
-    var redNodes = require(nrPath+"/red/runtime/nodes");
-    var flows = require(nrPath+"/red/runtime/nodes/flows");
-    var credentials = require(nrPath+"/red/runtime/nodes/credentials");
-    var comms = require(nrPath+"/red/api/editor/comms.js");
-    var log = require(nrPath+"/red/runtime/log.js");
-    var context = require(nrPath+"/red/runtime/nodes/context.js");
-    var events = require(nrPath+"/red/runtime/events.js");
+    // if not, we are running in the core
+    runtimePath = process.cwd()+"/red/red.js";
```

Comment:
There are more robust ways to determine this.  For example, you could use something like [read-pkg-up](https://www.npmjs.com/package/read-pkg-up) then test if the `name` of the result is `node-red`.

Reply:
That appending `.js` to a name in a `require()` call is not necessary, as it's assumed.  The only time this *is* necessary is if you have something like:

```
foo.js
foo/index.js
```

Then `require('foo')` does...what?  It's the same every time, but not obvious.

Reply:
Thanks - [read-pkg-up](https://www.npmjs.com/package/read-pkg-up) looks like a handy package.  Will see if I can rework things to use it.

Yes, I realize that, but the "main" of the node-red package is "red/red.js".  I need that module, as well as modules relative to "red" inside.  The path returned by `require.resolve('node-red')` is the full path to the red.js module (e.g. `/users/bob/my-node-package/node_modules/node-red/red/red.js` including the `red.js` at the end.  For consistency, I append `/red/red.js` to the current working directory if this fails, assuming I am running in the core.

Lines 52-57 get the path of the runtime if we are running in the core, or if node-red is a dependency.    This code does this synchronously when the helper is required so that existing core node tests need not call `init()`.

If I use [read-pkg-up](https://www.npmjs.com/package/read-pkg-up), I need to wait for results in a promise, so clients can't call helper methods until this is complete.  I will need to do this in the `load()` call I think...possibly others...

Anyway, I realize this is hacky, just need to take yet another step back...

Reply:
I see now that read-pkg-up can be called synchronously.  that makes things easier.